RHDEVDOCS-3631 Verify that builds work with RHEL subscriptions (#40993)
openshift/openshift-docs: modules/builds-running-entitled-builds-with-sharedsecret-objects.adoc

// Used in cicd/builds/running-entitled-builds.adoc

[id="builds-running-entitled-builds-with-sharedsecret-objects_{context}"]
= Running entitled builds with sharedsecret objects

[role="_abstract"]
You can run entitled builds with `SharedSecret` objects that import {op-system-base} Simple Content Access (SCA) certificates.  By using this feature, you can install entitled RPM packages during builds without copying your {op-system-base} subscription credentials and certificates into the builds' namespaces.

This procedure relies on the newly-introduced OpenShift Shared Resources feature and the Insights Operator.

[IMPORTANT]
====

> This enhancement is only available if you enable the `TechPreviewNoUpgrade` feature set. These Technology Preview features are not part of the default features. Enabling this feature set cannot be undone and prevents upgrades. This feature set is not recommended on production clusters. See xref:../post_installation_configuration/cluster-tasks.adoc#post-install-tp-tasks[Enabling Technology Preview features using FeatureGates].

Author: Excellent, @gabemontero! It's never too soon to get comments.

====

.Prerequisites

* You must xref:../post_installation_configuration/cluster-tasks.adoc#post-install-tp-tasks[enable Technology Preview features using FeatureGates].

.Procedure

> Under the [examples directory](examples) there is both a simple example of using a `SharedConfigMap` in a `Pod`, and an example of performing an OpenShift `Build` from a `BuildConfig` that uses Red Hat Entitlements available from your Red Hat Subscription when your subscription credentials are stored on your cluster via OpenShift Insights Operator, and resulting `Secret` for those credentials is made available to additional `Namespaces` via a `SharedSecret`.  Instructions for the `Pod` example are [here](docs/simple-example.md), and for the `BuildConfig` example are [here](docs/build-with-rhel-entitlements.md).

Reviewer: Looks like you grabbed the upstream markdown doc as is.

Certainly that should be pruned to just cite the BuildConfig one, and instead of going through the instructions that depend on contents of that repo, you should grab the yaml from the files in that repo and cite the yaml here in the doc, like we do with other examples in the build docs.

i.e. https://docs.openshift.com/container-platform/4.9/cicd/builds/running-entitled-builds.html

Now, as you have the WIP label on this PR, perhaps you have such changes in progress :-)

But it not, and you need further clarification / guidance, let's get on a video conf and discuss.

Author: I replaced this with the content from the gdoc for the blog post.

`openshift-config-managed` namespace, which is set up by the OpenShift Insights Operator if your install pull secret in fact has a valid
Red Hat subscription that entitles you to install Red Hat RPMs.  With a reference to this `SharedSecret`, the `BuildConfig`
is able to use your subscription information to install entitled RPMs when building an image.  Separate `Role` and
`RoleBinding` objects are defined for the `BuildConfig` and its use of the `SharedSecret`.

To install this example, run:

```shell
$ kubectl apply -R -f ./examples
```

You can start a `Build` for this `BuildConfig`, and follow the logs, via

```shell
$ oc start-build my-csi-bc -F
```

and watch as Red Hat entitled content via your subscription is installed during the running of the Docker Strategy build.
See the next section first however, if you have never used OpenShift Insights Operator to inject your subscription
credentials into your cluster.

## Notes or setting up your cluster so that Insights Operator injects your subscription credentials for use in your cluster

Tech Preview must be enabled on your OCP cluster to use the Shared Resources CSI Driver and to sync the Simple Content

> Access certificate of your Red Hat account.

Reviewer 2: The sync of the SCA will be done automatically by default in the 4.11. IOW the tech preview is not required.

Reviewer 3: The doc updated for 4.10+ should merge soon: https://github.com/openshift/openshift-docs/pull/40689 
That is the most up to date

Author: @gabemontero Should we make note of this in the 4.10 docs?

See [this link](https://docs.openshift.com/container-platform/4.9/post_installation_configuration/cluster-tasks.html#nodes-cluster-enabling-features-cli_post-install-cluster-tasks)
if you want to learn how to convert your cluster to a Tech Preview cluster.

Once that is sorted out, as noted up top, the Insights Operator will store your subscription credentials in the `Secret`
named `etc-pki-entitlement` in the `openshift-config-managed` namespace.  However, you'll need to perform some management

> actions around your subscription and your cluster to get the Insights Operator to actually create and maintain that `Secret`

Reviewer 2: Perhaps we could say what actions? Or at least link some other documentation?

Author: @gabemontero Should we make note of this in the 4.10 docs?

which will take you to entry for you cluster at (http://console.redhat.com)[http://console.redhat.com].

![link to console.redhat.com](ocp-console-ptr-cluster-manager.png)

Click that `OpenShift Cluster Manager` link and log in.  Once there, you should see a `Subscription settings` section,
with a link to edit your subscription settings.

![edit subscription pointer](cluster-manager-subscription-settings-ptr.png)

What you see after you click the `Edit subscription settings` link will vary based on the nature of your subscription.
Testing from the Shared Resource developers has shown you do have to select a choice beyond any sort of fixed term trial
in order to get your subscription to a place where Insights Operator will process it.  That said, "Self-support" is still
a valid option in order to access the entitled content available with your subscription.  Selecting that would look something like
this:

![subscription details](cluster-manager-subscription-settings.png)

Once you are at this point, wait for Insights Operator to pull in your credential into the `Secret`.  NOTE: Insights

> Operator does have a [polling interval](https://docs.openshift.com/container-platform/4.9/support/remote_health_monitoring/insights-operator-simple-access.html#insights-operator-configuring-sca_remote-health-reporting-from-restricted-network)

Reviewer 2: Shouldn't the link point to the latest docs please?

Reviewer 2: OK np. Thanks.

Author: @gabemontero Should we make note of this in the 4.10 docs?

// Used in cicd/builds/running-entitled-builds.adoc

[id="builds-running-entitled-builds-with-sharedsecret-objects_{context}"]
= Running entitled builds with SharedSecret objects

[role="_abstract"]
Rather than copying your {op-system-base} subscription credentials and certificates into a builds' namespace as a Secret, you can access your credentials and certificates from a `Secret` object in one of the {product-title} system namespaces. To access this `Secret` object, you run entitled builds with CSI volume mount of a `SharedSecret` custom resource (CR) instance that references the `Secret` object.

> This procedure relies on the newly-introduced Shared Resources feature and the Insights Operator in {product-title}.

Author: @gabemontero Is "Shared Resources" another name for the "Shared Resources CSI Driver"?

Reviewer: "Shared Resources" IMO based on context could be short hand for either the CSI Driver, the two new CRDs/API, or both the CSI Driver and the new CRDs/API we built (i.e. the "end to end solution"

In some places I use it as a short hand for "SharedSecrets and SharedConfigMaps".

Author: Updated. Thanks.

:FeatureName: The Shared Resources
include::snippets/technology-preview.adoc[leveloffset=+1]

[NOTE]
====
To enable the Shared Resources, you must enable features using feature gates. See "Enabling features using feature gates" in the following "Additional resources" section.
====

[IMPORTANT]
====

> Both the Build CSI Volume feature and Shared Resources feature are only available as Technology Preview features. They also belong to the `TechPreviewNoUpgrade` feature set, which is not part of the default features, and is not recommended on production clusters. You must enable the `TechPreviewNoUpgrade` feature set to use these features. Once enabled, you cannot disable the `TechPreviewNoUpgrade` feature set and cannot upgrade your cluster. See xref:../../post_installation_configuration/cluster-tasks.adoc#post-install-tp-tasks[Enabling Technology Preview features using feature gates].

Author: @gabemontero Let's talk about these two feature names on Monday. I want to make sure I use the correct/formal titles to help avoid confusing readers.

Reviewer: Sure.  To help set the table for that:
- Build CSI Volumes is BUILD-275 from @jkhelil which build on top of BUILD-87 from @coreydaley 
- While consuming Shared Resources is a key scenario for wanting BUILD-275, it is not the only one.  As @adambkaplan noted during planning, we have separate customer use cases where they want builds to utilize the other CSI Drviers that already exist (AWS EBS etc etc.)
- Build CSI Volume was introduced as a separate feature gate from CSIDriverSharedResources in the Tech preview features set by me with BUILD-281.  We did this because it is entirely possible we could promote one out of Tech preview before the other, based on how various security items.  But I don't think you need to dive into those specifics in the docs.

Author: Updated. Thanks.

For more information about the support scope of Red Hat Technology Preview
features, see https://access.redhat.com/support/offerings/techpreview/.

The Shared Resources CSI Driver and the Build CSI Volumes features also belong to the `TechPreviewNoUpgrade` feature set, which is a subset of the current Technology Preview features. You can enable the `TechPreviewNoUpgrade` feature set on test clusters, where you can fully test them while leaving the features disabled on production clusters. Enabling this feature set cannot be undone and prevents updates. This feature set is not recommended on production clusters. See "Enabling Technology Preview features using feature gates" in the following "Additional resources" section.
====

.Prerequisites

* You have enabled the  `TechPreviewNoUpgrade` feature set by using the feature gates.
* You have a `SharedSecret` custom resource (CR) instance that references the `Secret` object where the Insights Operator stores the subscription credentials.

> 

Reviewer 4: ```suggestion
```

// Used in cicd/builds/running-entitled-builds.adoc

[id="builds-running-entitled-builds-with-sharedsecret-objects_{context}"]
= Running entitled builds using SharedSecret objects

[role="_abstract"]
Configure and perform a build in one namespace that securely uses RHEL entitlements from a `Secret` object in another namespace.

> Previously, {product-title} 4.9.x and earlier suggested accessing RHEL entitlements from OpenShift Builds by creating a `Secret` object with your subscription credentials in the same namespace as your `Build` object. Now, with {product-title} 4.10 and later, you can access your credentials and certificates from a `Secret` object in one of the {product-title} system namespaces. You run entitled builds with a CSI volume mount of a `SharedSecret` custom resource (CR) instance that references the `Secret` object.

Reviewer 4: ```suggestion
Previously, in {product-title} 4.9.x and earlier, you could access RHEL entitlements from OpenShift Builds by creating a `Secret` object with your subscription credentials in the same namespace as your `Build` object. Now, with {product-title} 4.10 and later, you can access your credentials and certificates from a `Secret` object in one of the {product-title} system namespaces. You run entitled builds with a CSI volume mount of a `SharedSecret` custom resource (CR) instance that references the `Secret` object.
```
Avoid personifying the software.

> This procedure relies on the newly introduced The Shared Resources CSI Driver feature, which you can use to declare CSI Volume mounts in {product-title} Builds. It also relies on the {product-title} Insights Operator.

Reviewer 4: ```suggestion
This procedure relies on the newly introduced Shared Resources CSI Driver feature, which you can use to declare CSI Volume mounts in {product-title} Builds. It also relies on the {product-title} Insights Operator.
```

For more information about the support scope of Red Hat Technology Preview
features, see https://access.redhat.com/support/offerings/techpreview/.
====

[IMPORTANT]
====
The Shared Resources CSI Driver and The Build CSI Volumes features also belong to the `TechPreviewNoUpgrade` feature set, which is a subset of the current Technology Preview features. You can enable the `TechPreviewNoUpgrade` feature set on test clusters, where you can fully test them while leaving the features disabled on production clusters. Enabling this feature set cannot be undone and prevents updates. This feature set is not recommended on production clusters. See "Enabling Technology Preview features using feature gates" in the following "Additional resources" section.
====

.Prerequisites

* You have enabled the  `TechPreviewNoUpgrade` feature set by using the feature gates.
* You have a `SharedSecret` CR instance that references the `Secret` object where the Insights Operator stores the subscription credentials.
* You must have permission to perform the following actions

>   * Create pods.

Author: Also applied back to https://github.com/openshift/openshift-docs/pull/41898

>   * Grant the service acount of the pod permissions to use a Security Context Constraint (SCC) that allows Container Storage Interface (CSI) volumes to be mounted in a pod.

Author: Also applied back to https://github.com/openshift/openshift-docs/pull/41898

  * Discover which `SharedSecret` CR instances are available by entering the `oc get sharedsecrets` command and getting a non-empty list back.

>   * Determine if any of the service accounts available to you in your namespace are allowed to use a given `SharedSecret` CR instance. In other words, you can run `oc adm policy who-can use <identifier of specific SharedSecret>` to see if a service account in your namespace is listed.

Author: Also applied back to https://github.com/openshift/openshift-docs/pull/41898

[NOTE]
====
If neither of the last two prerequisites in this list are met, establish, or ask someone to establish, the necessary role-based access control (RBAC) so that you can discover `SharedSecret` CR instances and enable service accounts to use `SharedSecret` CR instances.
====

.Procedure

> . Grant a given service account RBAC permissions to use the `SharedSecret` CR instance in its pod by using `oc apply` with YAML content:

Author: Also applied back to https://github.com/openshift/openshift-docs/pull/41898

+
[NOTE]
====

> Currently, `kubectl` and `oc` have hard-coded special case logic restricting the `use` verb to roles centered around pod security. Therefore, you cannot use `oc create role ...` to create the role needed for consuming `SharedSecret` CR instance.

Reviewer 4: ```suggestion
Currently, `kubectl` and `oc` have hard-coded special case logic restricting the `use` verb to roles centered around pod security. Therefore, you cannot use `oc create role ...` to create the role needed for consuming `SharedSecret` CR instances.
```

====
+
[source,terminal]
----
$ oc apply -f - <<EOF
apiVersion: rbac.authorization.k8s.io/v1
kind: Role
metadata:
  name: shared-resource-my-share
  namespace: my-namespace
rules:
  - apiGroups:
      - sharedresource.openshift.io
    resources:
      - sharedsecrets
    resourceNames:
      - my-share
    verbs:
      - use
EOF
----

. Create the `RoleBinding` associated with the role by using the `oc` command:
+
[source,terminal]
----

> $ oc create rolebinding shared-resource-my-share --role=shared-resource-my-share --serviceaccount=my-namespace:default

Author: Also applied back to https://github.com/openshift/openshift-docs/pull/41898

----

> . Access the `SharedSecret` CR instance from a pod:

Author: ?

= Running entitled builds using SharedSecret objects

[role="_abstract"]
Configure and perform a build in one namespace that securely uses RHEL entitlements from a `Secret` object in another namespace.

Previously, {product-title} 4.9.x and earlier suggested accessing RHEL entitlements from OpenShift Builds by creating a `Secret` object with your subscription credentials in the same namespace as your `Build` object. Now, with {product-title} 4.10 and later, you can access your credentials and certificates from a `Secret` object in one of the {product-title} system namespaces. You run entitled builds with CSI volume mount of a `SharedSecret` custom resource (CR) instance that references the `Secret` object.

This procedure relies on the newly introduced The Shared Resources CSI Driver feature, which you can use to declare CSI Volume mounts in {product-title} Builds. It also relies on the {product-title} Insights Operator.

:FeatureName: The Shared Resources CSI Driver
include::snippets/technology-preview.adoc[leveloffset=+1]

:FeatureName: TBD
include::snippets/technology-preview.adoc[leveloffset=+1]

[IMPORTANT]
====

> The Shared Resources CSI Driver and TBD features also belong to the `TechPreviewNoUpgrade` feature set, which is a subset of the current Technology Preview features. You can enable the `TechPreviewNoUpgrade` feature set on test clusters, where you can fully test them while leaving the features disabled on production clusters. Enabling this feature set cannot be undone and prevents updates. This feature set is not recommended on production clusters. See "Enabling Technology Preview features using feature gates" in the following "Additional resources" section.

Reviewer: Looking at https://deploy-preview-40993--osdocs.netlify.app/openshift-enterprise/latest/cicd/builds/running-entitled-builds.html#builds-running-entitled-builds-with-sharedsecret-objects_running-entitled-builds I think the two included items above are sufficient, and this extra item here is redundant and should be removed

====

> .User roles

Reviewer: Yeah for the blog posts clarifying the user roles was important given the differing audience, but here in the official doc, it is unecessary.

You should make it more generic like we have in the shared resource doc.

So remove the user role section.

Then in the procedure sction start with "You need to have permission to do the following .." or something like that and then move on to the numbered list there.

The following procedure includes several types of users:

* An _administrator_ with cluster-level permissions to access multiple namespaces, create cluster-scoped objects, and create `Role` and `RoleBinding` objects around those cluster-scoped objects, both for discovery and use.
* Optionally, _namespace administrators_ can perform some of the tasks in this example, such as creating  `Role` and `RoleBinding` objects for users or service accounts.
* General users of the cluster, such as  _developers_ or _testers_, are generally restricted to access a single namespace to build images with OpenShift Builds.
+
So that these users can use RHEL entitlements in their builds, administrators give them access to discover `SharedSecret` and `SharedConfigMaps` custom resource (CR) instances that they can reference in their `BuildConfig` objects. Also, their namespace `builder` service account has been permitted to mount the specific `SharedSecret` CR instance as a Shared Resource CSI Driver volume in OpenShift Build pods.

.Prerequisites

* You have enabled the  `TechPreviewNoUpgrade` feature set by using the feature gates.

> * You have created and registered a `Secret` object at https://console.redhat.com/.

Reviewer: Yeah the blog post has 

```
A SharedSecret that references the Secret where the Insights Operator stores the Subscription Credentials available to the cluster created and registered at https://console.redhat.com/
```

Did you try to "simplify" things?  If so, you've gone to the point where what you have here is inaccurate / misleading.

.Procedure

. As a cluster or namespace administrator, create a `Namespace` object to host your `BuildConfig` and `Build` objects.
+
.Example YAML Namespace object definition
[source,yaml]
----
apiVersion: v1
kind: Namespace
metadata:
  name: my-csi-app-namespace
----

. As a cluster administrator, create a `SharedSecret` custom resource (CR) instance that references the `Secret` object where the Insights Operator stores the subscription credentials for the cluster.
+
.Example YAML `SharedSecret` custom resource definition
[source,yaml]
----
apiVersion: sharedresource.openshift.io/v1alpha1
kind: SharedSecret
metadata:
  name: my-share-bc
spec:
  secretRef:
    name: etc-pki-entitlement
    namespace: openshift-config-managed
----

. As a cluster or namespace administrator, create a `Role` object so that the `builder` service account in our namespace can access the `SharedSecret` CR instance.
+

> .Example YAML `Role` object definition

Reviewer: can you add the tidbit we in the shared resource section where oc/kubectl won't let us do `oc create role`

[source,yaml]
----
apiVersion: rbac.authorization.k8s.io/v1
kind: Role
metadata:
  name: shared-resource-my-share-bc
  namespace: my-csi-app-namespace
rules:
  - apiGroups:
      - sharedresource.openshift.io
    resources:
      - sharedsecrets
    resourceNames:
      - my-share-bc
    verbs:
      - use
----

. As a cluster or namespace administrator, create a `RoleBinding` object so that the `builder` service account in our namespace can access the `SharedSecret` CR instance.
+

> .Example YAML `RoleBinding` object definition

Reviewer: convert this to an `oc create rolebinding...` like in the shared secret doc

just name sure you don't just copy/paste as the names of the role need to match the -bc suffix we have here.

  kind: Role
  name: shared-resource-my-share-bc
subjects:
  - kind: ServiceAccount
    name: builder
    namespace: my-csi-app-namespace
----

. As a developer or tester in a given namespace, create a `BuildConfig` object that accesses the RHEL entitlements.
+
.Example YAML `BuildConfig` object definition
[source,yaml]
----
apiVersion: build.openshift.io/v1
kind: BuildConfig
metadata:
  name: my-csi-bc
  namespace: my-csi-app-namespace
spec:
  runPolicy: Serial
  source:
    dockerfile: |
      FROM registry.redhat.io/ubi8/ubi:latest
      RUN ls -la /etc/pki/entitlement
      RUN rm /etc/rhsm-host
      RUN yum repolist --disablerepo=*
      RUN subscription-manager repos --enable rhocp-4.9-for-rhel-8-x86_64-rpms
      RUN yum -y update
      RUN yum install -y openshift-clients.x86_64
  strategy:
    type: Docker
    dockerStrategy:
      volumes:
        - mounts:
            - destinationPath: "/etc/pki/entitlement"
          name: my-csi-shared-secret
          source:
            csi:
              driver: csi.sharedresource.openshift.io
              readOnly: true
              volumeAttributes:
                sharedSecret: my-share-bc
            type: CSI
----

> . As a developer or tester, start a build from the `BuildConfig` object and follow the logs with the `oc` command.

Reviewer: again lose the "user role" reference.

+
.Example oc start-build command
[source,terminal]
----
$ oc start-build my-csi-bc -F
----
+

> .Example output from the oc start-build command

Reviewer: mention the blub that we have filtered some of the output with "..." etc.

// Used in cicd/builds/running-entitled-builds.adoc

[id="builds-running-entitled-builds-with-sharedsecret-objects_{context}"]
= Running entitled builds with SharedSecret objects

[role="_abstract"]

> To avoid copying your {op-system-base} subscription credentials and certificates into the builds' namespaces, you can install entitled RPM packages during builds. To install entitled RPMs, you run entitled builds with `SharedSecret` objects that import {op-system-base} Simple Content Access (SCA) certificates.

Author: Updated. Thanks.

> This procedure relies on the newly-introduced Shared Resources feature and the Insights Operator in {product-title}.

Reviewer: ```suggestion
This procedure relies on the newly-introduced Shared Resources feature, the ability to declare CSI Volume mounts in {product-title} Builds, and the Insights Operator in {product-title}.
```

Author: Updated. Thanks.

[IMPORTANT]
====

> The Shared Resources feature is only available as a Technology Preview feature. It also belongs to the `TechPreviewNoUpgrade` feature set, which are not part of the default features, and are not recommended on production clusters. To use the Shared Resources feature, you must enable the `TechPreviewNoUpgrade` feature set. Once enabled, you cannot disable the `TechPreviewNoUpgrade` feature set and cannot upgrade your cluster. See xref:../../post_installation_configuration/cluster-tasks.adoc#post-install-tp-tasks[Enabling Technology Preview features using FeatureGates].

Reviewer: ```suggestion
Both the Build CSI Volume feature and Shared Resources feature are only available as a Technology Preview feature. It also belongs to the `TechPreviewNoUpgrade` feature set, which are not part of the default features, and are not recommended on production clusters. To use these features, you must enable the `TechPreviewNoUpgrade` feature set. Once enabled, you cannot disable the `TechPreviewNoUpgrade` feature set and cannot upgrade your cluster. See xref:../../post_installation_configuration/cluster-tasks.adoc#post-install-tp-tasks[Enabling Technology Preview features using FeatureGates].
```

Author: Updated. Thanks.

====

.Prerequisites

* You have xref:../../post_installation_configuration/cluster-tasks.adoc#post-install-tp-tasks[enabled the  `TechPreviewNoUpgrade` feature set by using the FeatureGates].

.Procedure

> Under the [examples directory](examples) there is both a simple example of using a `SharedConfigMap` in a `Pod`, and an

Reviewer: Maybe you already had this planned since this is a WIP, but it not, sorry @rolfedh but the entire Procedure section here needs to be changed to *NOT* use direct references to the examples dir in our source code repo.

I *think* you are planning to borro from the blow post draft, right?

If so, yes, a filtered down version of the "What can we demonstrate" section https://docs.google.com/document/d/1SACY6vX0IaxIUr4cL6g1tqLDVM7Oh3NHi5DbcwIu8Vk/edit#heading=h.220lhy819rda of the blog post is what is needed.

Do you feel comfortable doing that filtering / pruning / paring down, or do you need me to take a stab at it?

Author: This is excellent feedback, @gabemontero. I appreciate it. I'll be moving this content over today so that you can review it early on Monday. I'll incorporate your changes and try to get this PR merged by EOD Monday.

Author: Updated. Thanks. Ready for your review.
